File: 234-palindrome-linked-list/palindrome-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* slow=head;
        ListNode* fast=head;
        while(fast!=nullptr && fast->next!=nullptr)
        {
            slow=slow->next;
            fast=fast->next->next;
        }
        ListNode* prev=nullptr;
        ListNode* curr=slow;
        while(curr!=nullptr)
        {
            ListNode* temp=curr->next;
            curr->next=prev;
            prev=curr;
            curr=temp;
        }
       ListNode* firsthalf=head;
       ListNode* secondhalf=prev;
       while(secondhalf!=nullptr)
       {
        if(firsthalf->val!=secondhalf->val)
        {
            return false;
        }
        firsthalf=firsthalf->next;
        secondhalf=secondhalf->next;
       } 
       return true;
    }
};

File: 234-palindrome-linked-list/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow=head
        fast=head
        while(fast!=None and fast.next!=None):
            slow=slow.next
            fast=fast.next.next
        prev=None
        curr=slow
        while(curr!=None):
            temp=curr.next
            curr.next=prev
            prev=curr
            curr=temp
        firstlist=head
        secondlist=prev
        while(secondlist!=None):
            if(firstlist.val!=secondlist.val):
                return False
            firstlist=firstlist.next
            secondlist=secondlist.next
        return True


        